testing/test_dataset.py

```diff
@@ -1424,6 +1517,66 @@ def testArrayMinMeanMax(self):
         # Run tests
         np.testing.assert_array_equal(result, expected)
 
+    def checkMMMShape(self, MMMArrs: dict, expMMMShapes: dict):
+        """ Utility method to check that the given dict of MMMs has the right shape """
+        accMMMShapes = {chID: np.shape(MMMArr) for chID, MMMArr in MMMArrs.items()}
+        try:
+            assert accMMMShapes == expMMMShapes
+        except AssertionError as e:
+            return [str(e)]
+        return []
+
+    def checkMMMComparisons(self, MMMArrs: dict):
+        """ Utility method to check that mins < means < maxes """
+        failedAsserts = []
+        for chID, MMMArr in MMMArrs.items():
+            if np.greater(MMMArr[0][1:], MMMArr[1][1:]).any():
+                failedAsserts.append(f"CH{chID} FAILED MIN < MEAN\nMIN:\n{MMMArr[0][1:]}\nMEAN:\n{MMMArr[1][1:]}")
+            if np.greater(MMMArr[0][1:], MMMArr[2][1:]).any():
+                failedAsserts.append(f"CH{chID} FAILED MEAN < MAX\nMEAN:\n{MMMArr[1][1:]}\nMAX:\n{MMMArr[2][1:]}")
+
+        return failedAsserts
+
+    @pytest.mark.parametrize()
+    def checkValues(self, MMMArrs: dict, expBlock0s: dict):
+        """ Utility method to check the first block's values in all provided arrayMinMeanMaxes """
+
+        failedAsserts = []
+        for ch, MMMArr in expBlock0s.items():
+            accBlock0 = MMMArrs[ch][:, 1:, 0]
+            if np.not_equal(accBlock0, expBlock0s[ch]).any():
+                failedAsserts.append(f"Channel {ch} MMM Block 0 Expected:\n{expBlock0s[ch]}\nGot:\n{accBlock0}")
+
+        return failedAsserts
+
+    @pytest.mark.parametrize('channels', [(8, 80), (84, 20, 59, 76)], ids=['Original MMM Encoded', 'Original Non-encoded'])
+    def testS5E25D40MMM(self, S5E25D40IDE, S5E25D40MMMs, S5E25D40Shapes, S5E25D40Block0s, channels):
+        MMMs = {chID: MMM for chID, MMM in S5E25D40MMMs.items() if chID in channels}
+        failedAsserts = self.checkMMMComparisons(MMMs)
+        failedAsserts += self.checkMMMShape(S5E25D40MMMs, S5E25D40Shapes)
+        assert not failedAsserts, "\n".join(failedAsserts)
+
+        # keep only the first block values for the channels that are in MMMs/being tested
+        block0s = {channel: block0 for channel, block0 in S5E25D40Block0s.items() if channel in channels}
+        failedAsserts += self.checkValues(S5E25D40MMMs, block0s)
+
+        # raise one big failure to show all inner failures
+        assert not failedAsserts, "\n".join(failedAsserts)
```

Comment:
the inner self.checkXYZ methods allow for a larger list of all the errors rather than breaking at the first assert that fails. Should I change this? The process is that shape is checked first, then min<mean<max, then the values themselves

Reply:
I think it would be better to have each test generate one exception. If you want multiple checks to run even if one failed, make them separate, smaller tests. Development tools may have difficulty parsing the joined lists of exceptions.